# Wrapper for the aligned_perc_calc_functions.r script

# Load packages
suppressPackageStartupMessages(library(Biostrings))
suppressPackageStartupMessages(library(argparse))


parser <- ArgumentParser(description='Calculate percentage of aligned bases')
parser$add_argument("-f", "--filename", metavar = "character", type = "character", nargs="+", help = "list of input files")
parser$add_argument("--fasta", metavar = "character", type = "character", nargs="+", help = "list of query fasta files")
parser$add_argument("--nfile", metavar = "character", type = "character", help = "the file generated by get_N_locations")
parser$add_argument("--out", metavar = "character", type = "character", nargs="+", help = "list of output files")
parser$add_argument("-l", "--length_threshold", metavar = "integer", type = "integer", default =  1000, help = "threshold value for length filter [default]")
parser$add_argument("-i", "--identity_threshold", metavar = "double", type = "double", default = 90.0, help = "threshold value for identity filter [default]")
#parser$add_argument("--n_threshold_abs", metavar = "integer", type = "integer", default = 1000, help = "threshold value for maximum number of N nucleotides in alignment [default]")
parser$add_argument("--n_threshold_perc", metavar = "double", type = "double", default = 10.0, help = "threshold value for maximum percentage of N in alignment [default]")

opt <- parser$parse_args()

# Load functions
source("scripts/aligned_perc_calc_functions.r")

mode <- 1 # mode 1 is wrapper mode, 0 is standalone mode
# remake of aligned_perc_calc.py
oldtime <- Sys.time() # test speed of script
coords_files <- opt$filename
for (i in coords_files){
  filtered_data <- get_filtered_data(filename = i,
                                     length_threshold = opt$length_threshold,
                                     identity_threshold = opt$identity_threshold,
                                     mode = 1) # filter the data
  merged <- merge_data(filename = gsub(".coords",".tsv",gsub("sorted","Temp/filtered",i)),mode = 1, data = filtered_data) # remove redundant sequences and merge overlapping sequences
  N_filtered_data <- filter_N_entries(data = merged, N_file = opt$nfile, N_threshold_perc = opt$n_threshold_perc)
  get_aligned_perc(filename = gsub(".coords",".NR.tsv",gsub("sorted","Temp/filtered",i)),mode = 1, data = merged, fastafile = opt$fasta, out = opt$out) # calculate the percentage of aligned bases
}
#print(Sys.time()-oldtime)
