# filtering script of the snakemake mummer pipeline (by Bleeker group)

suppressPackageStartupMessages(library(Biostrings))
suppressPackageStartupMessages(library(argparse))


parser <- ArgumentParser(description='Calculate percentage of aligned bases')
parser$add_argument("-f", "--filename", metavar = "character", type = "character", help = "list of input files")
#parser$add_argument("--fasta", metavar = "character", type = "character", nargs="+", help = "list of query fasta files")
parser$add_argument("--nfile", metavar = "character", type = "character", help = "the file generated by get_N_locations")
parser$add_argument("--out", metavar = "character", type = "character", nargs="+", help = "list of output files")
parser$add_argument("-l", "--length_threshold", metavar = "integer", type = "integer", default =  1000, help = "threshold value for length filter [default]")
parser$add_argument("-i", "--identity_threshold", metavar = "double", type = "double", default = 90.0, help = "threshold value for identity filter [default]")
parser$add_argument("--n_threshold_perc", metavar = "double", type = "double", default = 10.0, help = "threshold value for maximum percentage of N in alignment [default]")

opt <- parser$parse_args()

# Load functions
source("scripts/aligned_perc_calc_functions.r")

# filter the alignments based on length and identity
filtered_data <- get_filtered_data(filename = opt$filename,
                                   length_threshold = opt$length_threshold,
                                   identity_threshold = opt$identity_threshold)

# filter the alignments based on percentage of n nucleotides
filter_N_entries(data = filtered_data, N_file = opt$nfile, N_threshold_perc = opt$n_threshold_perc, mode = 0, outname = opt$out)

#merged <- merge_data(filename = gsub(".coords",".tsv",gsub("sorted","Temp/filtered",i)),mode = 1, data = N_filtered_data) # remove redundant sequences and merge overlapping sequences
#get_aligned_perc(filename = gsub(".coords",".NR.tsv",gsub("sorted","Temp/filtered",i)),mode = 1, data = merged, fastafile = opt$fasta, out = opt$out) # calculate the percentage of aligned bases
